### PR TITLE
Modify the mini batch inner loop in the _plain_sgd function

### PR DESCRIPTION
**Pull request by VishwaP98**

### CODE REVIEW

File: sklearn/linear_model/_sgd_fast.pyx [hunk 1]
```diff
@@ -545,61 +545,67 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
...
+
+                # also calculate the loss and dloss
+                # l = loss.loss(p, mini_batch_y)
+
+                #if verbose or not early_stopping:
+                #    sumloss += loss.loss(p, mini_batch_y)
+
+                avg_batch_loss = sum_batch_loss / batch_size
+                avg_batch_dloss = sum_batch_dloss / batch_size
+                batch_x_sum.scale(1/batch_size)
```
Comment (RylandWang) [inline sklearn/linear_model/_sgd_fast.pyx:607]: what's the purpose of batch_x_sum?

File: sklearn/linear_model/_sgd_fast.pyx [hunk 2]
```diff
@@ -621,7 +627,7 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
                         # regression
                         update *= -1
 
-                update *= class_weight * sample_weight
+                # update *= class_weight * sample_weight
```
Comment (RylandWang) [inline sklearn/linear_model/_sgd_fast.pyx:630]: My understanding is that this calculation determines whether the update is positive or negative.

Comment (RylandWang) [reply sklearn/linear_model/_sgd_fast.pyx:630]: We should preserve it even if we don't fully understand it.

**Review by RylandWang — CHANGES_REQUESTED** [review]
I see a couple of compile-time and runtime errors in your code. Could you fix these and test your code?

File: sklearn/linear_model/_sgd_fast.pyx [hunk 3]
```diff
@@ -457,6 +457,10 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
     cdef Py_ssize_t n_features = weights.shape[0]
 
     cdef WeightVector w = WeightVector(weights, average_weights)
+    cdef WeightVector mini_batch_x = WeightVector(np.zeros(batch_size), None)
+    cdef double mini_batch_y[batch_size]
+    cdef float mini_class_weights[batch_size]
+    cdef double sample_weight
```
Comment (linjun9) [inline sklearn/linear_model/_sgd_fast.pyx:463]: can you put this one back?

File: sklearn/linear_model/_sgd_fast.pyx [hunk 4]
```diff
@@ -457,6 +457,10 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
     cdef Py_ssize_t n_features = weights.shape[0]
 
     cdef WeightVector w = WeightVector(weights, average_weights)
+    cdef WeightVector mini_batch_x = WeightVector(np.zeros(batch_size), None)
+    cdef double mini_batch_y[batch_size]
+    cdef float mini_class_weights[batch_size]
```
Comment (linjun9) [inline sklearn/linear_model/_sgd_fast.pyx:462]: why float? I think `double` is better.

File: sklearn/linear_model/_sgd_fast.pyx (hunk 1 above)
Comment (linjun9) [reply sklearn/linear_model/_sgd_fast.pyx:607]: yea, we should delete this line. This was my code before, I was using `batch_x_sum` to add up all the x values. Now Vishwa declared a new variable `mini_batch_x`. I think they are doing the same thing. maybe better naming?

File: sklearn/utils/_weight_vector.pyx [hunk 5]
```diff
@@ -42,7 +42,7 @@ cdef class WeightVector(object):
         The squared norm of ``w``.
     """
     def __cinit__(self, double [::1] w, double [::1] aw):
-        if w.shape[0] > INT_MAX:
+        if w.shape[0] > INT_M AX:
```
Comment (linjun9) [inline sklearn/utils/_weight_vector.pyx:45]: mistake here.

File: sklearn/linear_model/_sgd_fast.pyx [hunk 6]
```diff
@@ -636,8 +642,9 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
                 # TODO: not sure about this part
                 if update != 0.0:
                     # TODO: add a for loop, loop through the  batch_data_pointers , batch_index_pointers, xnnzs,
-                    # update the w
-                    w.add(x_data_ptr, x_ind_ptr, xnnz, update)
+                    # update all the items we have in our mini size batch with the update
+                    for index in range(len(batch_size)):
```
Comment (linjun9) [inline sklearn/linear_model/_sgd_fast.pyx:646]: this is wrong. we should update w by `w.add() for all the x_data saved in the mini_batch_x`.

File: sklearn/linear_model/_sgd_fast.pyx (hunk 1 above)
Comment (linjun9) [reply sklearn/linear_model/_sgd_fast.pyx:607]: Also, you cannot save the all the pointers in a WeightVector `mini_batch_x`. You won't be able to retrieve the data pointers out.

**Pull request closed without merging**